### PR TITLE
Handle transposes in second batch of matrices in bmm

### PR DESCRIPTION
**Pull request by kulinseth**
1. Add support for Unranked placeholders in the MPS backend. 
2. PR is now fusing the Transposes into the GEMM kernel dispatches in MPS backend. This improves the performance of Transformer networks by 5-8%.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/LinearAlgebra.mm [hunk 1]
```diff
@@ -430,6 +430,26 @@ bool use_metal_mm(const Tensor& self, const Tensor& other, const Tensor& output)
...
+
+  // Handle transposes for the second batch of matrices.
+  if (batch2.is_view() && !batch2.is_contiguous()) {
+    if (batch2.numel() == batch2._base().numel()) {
+      const IntArrayRef& viewSizes = batch2.sizes();
+
+      // Handle 3D and 4D tensors.
+      // For 4D tensors, first it must have been reshaped from 4D to 3D and then transposed.
+      int32_t baseTransposeStrideDim = batch2._base().dim() == 4 ? -3 : -2;
+      if (batch2.size(0) == batch2.size(0) &&
```
Comment (DenisVieriu97) [inline aten/src/ATen/native/mps/operations/LinearAlgebra.mm:444]: this first check is not needed

**Review by DenisVieriu97 — APPROVED** [review]
Looks good - added a comment. Please take a look

**Comment by kit1980** [comment]
> @pytorchbot merge -f "All MPS tests are green"

Please be more careful next time, as this PR broke lint.

**Review by malfet — CHANGES_REQUESTED** [review]
Can you please provide a bit more detailed description of what this PR is trying to do. Is it fixing correctness issue or a performance one?

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 2]
```diff
@@ -276,26 +276,29 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
...
+    // The key format per tensor would look like ":Float32[1,1,1,10]:"
+    for (const Tensor& tensor: tensors) {
+      str += ":";
+      if (tensor.defined()) {
+        str += getMPSTypeString(tensor.scalar_type(), short_dtype) + "[";
+        // if tensor is a scalar
+        if (tensor.dim() == 0) {
+          str += "Scalar";
+        } else {
+          if (exclude_shape) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/OperationUtils.mm:290]: Why would we ever want to exclude shape from the TensorKey?

Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/OperationUtils.mm:290]: MPS supports unranked placeholders for which we don't need to know the shape to compile the graph - for cases where the graph compilation is very high due to shapes changing very frequently, we can make these ops unranked directly and exclude the shape from the key

Comment (malfet) [reply aten/src/ATen/native/mps/OperationUtils.mm:290]: @DenisVieriu97 Do you mind making this change as a separate PR? Are unranked placeholders supported for all ops or just matmuls? What are performance implications of this one?

**Comment by DenisVieriu97** [comment]
> Can you please provide a bit more detailed description of what this PR is trying to do. Is it fixing correctness issue or a performance one?

This improves Batched matrix multiplications where the 2nd matrix is transposed. E.g consider following example:

```
x = torch.randn(3, 200, 100, device="mps")
y = torch.randn(3, 500, 100, device="mps")

res = x @ y.permute(0, 2, 1)
```
With current code, a gather would be performed to make following `y.permute(0, 2, 1)` contiguous, then a batched matrix multiplication would be performed (where the matmul is performed `row @ column`).
With this addition,  we detect if the 2nd matrix was transposed, and if this was the case, we will not do gather anymore to make the 2nd matrix contiguous, but rather do directly a matrix multiplication where 2nd matrix is marked as transposed in MPSGraph. The big speedup comes that instead of being a row @ column matmul, now is a row @ row matmul. The big speedup in `row @ row` comes from the fact that the access to the 2nd matrix is not strided anymore (in trying to access the column), but rather contiguous.

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 2 above)
Comment (kulinseth) [reply aten/src/ATen/native/mps/OperationUtils.mm:290]: @malfet, I think these are 2 simple changes which can go together. The one is the change for unranked placeholders and its benefit now.

Comment (malfet) [reply aten/src/ATen/native/mps/OperationUtils.mm:290]: @kulinseth the risk here is that if it causes some regressions it had to be rolled back together

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 3]
```diff
@@ -28,6 +29,56 @@
...
+  NSError *error = nil;
+  id<MTLLibrary> library = compileCopyCastOpsLibrary(device, dtypeSrc, dtypeDst);
+  id<MTLFunction> func = [library newFunctionWithName:[NSString stringWithUTF8String:kernel.c_str()]];
+  TORCH_CHECK(func, "Failed to load the Metal Shader function: ", kernel);
+  id<MTLComputePipelineState> pso = [device newComputePipelineStateWithFunction:func error:&error];
+  TORCH_CHECK(pso != nil && error == nil, "Failed to construct pipeline state: ", [[error localizedDescription] UTF8String]);
+  _mtlPipelineCache[key] = pso;
+  return pso;
+}
+
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Copy.mm:81]: This isn't used anywhere, and seems unrelated to PR title, so deleting
```suggestion
```

File: test/test_mps.py [hunk 4]
```diff
@@ -412,6 +412,7 @@ def mps_ops_modifier(ops):
         'nn.functional.pixel_shuffle',
         'nn.functional.pixel_unshuffle',
         'nn.functional.tanhshrink',
+        'nn.functional.rms_norm',
```
Comment (malfet) [inline test/test_mps.py:415]: Not needed anymore
```suggestion
```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 5]
```diff
@@ -11,6 +11,7 @@
 #include <ATen/ops/real.h>
 #include <ATen/ops/view_as_real.h>
 #include <ATen/ops/zeros_like.h>
+#include <fmt/format.h>
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Copy.mm:14]: This one aren't needed either
```suggestion
```
